draft-ietf-tls-tls13.md

```diff
@@ -2810,32 +2827,82 @@ configuration identifier for each connection, thus preventing
 replay. Implementations are responsible for ensuring uniqueness of the
 identifier in this case.
 
+The "ticket_age" extension sent by the client SHOULD be used by
+servers to limit the time over which the first flight might be
+replayed.  A server can store the time at which it sends a server
+configuration to a client, or encode the time in a ticket.  Then, each
+time it receives an early_data extension, it can check to see if the
+value used by the client matches its expectations.
+
+The "ticket_age" value provided by the client will be shorter than the
+actual time elapsed on the server by a single round trip time.  This
+difference is comprised of the delay in sending the NewSessionTicket
+message to the client, plus the time taken to send the ClientHello to
+the server.  For this reason, a server SHOULD measure the round trip
+time prior to sending the NewSessionTicket message and account for
+that in the value it saves.
+
+There are several potential sources of error that make an exact
+measurement of time difficult.  Variations in client and server clocks
+are likely to be minimal, outside of gross time corrections.  Network
+propagation delays are most likely causes of a mismatch in legitimate
+values for elapsed time.  Both the NewSessionTicket and ClientHello
+messages might be retransmitted and therefore delayed, which might be
+hidden by TCP.
+
+A small allowance for errors in clocks and variations in measurements
+is advisable.  However, any allowance also increases the opportunity
+for replay.  In this case, it is better to reject early data than to
+risk greater exposure to replay attacks.
+
+#### Ticket Age
+
+%%% Key Exchange Messages
+       struct {
+           uint64 ticket_age;
+       } TicketAge;
+
+When the client sends the "early_data" extension, it MUST also send
+a "ticket_age" extension in its EncryptedExtensions block. This value
+contains the time elapsed since the client learned about the server
+configuration that it is using, in milliseconds.  This value can
+be used by the server to limit the time over which early data can
+be replayed.
```

Comment:
We need to distinguish between two different kinds of limits:

1. How large this value can be.
2. How large the *delta* between this value and the age of the ticket can be.

Any limits on how large the age of the ticket can be belong in NewSessionTicket, not here. With that said, since NST is limited to 86400 seconds, it seems like we could make this uint32.

WRT to how much delta/error is allowed, I'd rather leave that to the discretion of the server, as it seems like it's very domain specific. For instance, if you have a single machine with an anti-replay cache and restarts take 2 minutes, then multiple seconds would be reasonabl